index.d.ts

```diff
@@ -1,184 +1,289 @@
-import { Dispatch, SetStateAction, ReactNode } from 'react';
-
 declare module 'react-table' {
-    export interface Cell<D> {
-        render: (type: string) => any;
-        getCellProps: () => any;
-        column: Column<D>;
-        row: Row<D>;
-        state: any;
-        value: any;
-    }
+  import { ReactNode, useState } from 'react'
 
-    export interface Row<D> {
-        index: number;
-        cells: Cell<D>[];
-        getRowProps: () => any;
-        original: any;
-    }
+  type StringKey<D> = Extract<keyof D, string>
 
-    export interface HeaderColumn<D, A extends keyof D = never> {
-        /**
-         * This string/function is used to build the data model for your column.
-         */
-        accessor: A | ((originalRow: D) => string);
-        Header?: string | ((props: TableInstance<D>) => ReactNode);
-        Filter?: string | ((props: TableInstance<D>) => ReactNode);
-        Cell?: string | ((cell: Cell<D>) => ReactNode);
-
-        /**
-         * This is the unique ID for the column. It is used by reference in things like sorting, grouping, filtering etc.
-         */
-        id?: string | number;
-        minWidth?: string | number;
-        maxWidth?: string | number;
-        width?: string | number;
-        canSortBy?: boolean;
-        sortByFn?: (a: any, b: any, desc: boolean) => 0 | 1 | -1;
-        defaultSortDesc?: boolean;
-    }
+  type SortingRule<D, K extends StringKey<D> = StringKey<D>> = {
+    id: K
+    desc: boolean
+  }
 
-    export interface Column<D, A extends keyof D = never> extends HeaderColumn<D, A> {
-        id: string | number;
-    }
+  export type SortingRules<D> = SortingRule<D>[]
 
-    export type Page<D> = Row<D>[];
+  export type SortByFn = (a: any, b: any, desc: boolean) => 0 | 1 | -1
 
-    export interface EnhancedColumn<D, A extends keyof D = never> extends Column<D, A> {
-        render: (type: string) => any;
-        getHeaderProps: (userProps?: any) => any;
-        getSortByToggleProps: (userProps?: any) => any;
-        sorted: boolean;
-        sortedDesc: boolean;
-        sortedIndex: number;
-    }
+  export type Filters<D> = {
```

Comment:
I think this is the only thing left that I noticed. This can be `Record<StringKey<D>, string>`

Reply:
actually this didn't quite work since the key for the filter should be anything that can be an ID (same with sort), and that can include string.

Either way, I pulled out an IdType and then used that.

I'm still not particularly keen on allowing string in that type, it pretty much removes most of the explicit type checking we're jumping through hoops to get, but I also don't see how to avoid it.  It's pretty pervasive.